agent/config/config_accessor.go

```diff
@@ -0,0 +1,78 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package config
+
+// agentConfigAccessor struct implements AgentConfigAccessor interface defined in ecs-agent module.
+type agentConfigAccessor struct {
+	cfg *Config
+}
+
+// NewAgentConfigAccessor creates a new agentConfigAccessor.
+func NewAgentConfigAccessor(cfg *Config) *agentConfigAccessor {
+	return &agentConfigAccessor{
+		cfg: cfg,
+	}
+}
+
+func (aca agentConfigAccessor) APIEndpoint() string {
+	return aca.cfg.APIEndpoint
+}
+
+func (aca agentConfigAccessor) AWSRegion() string {
+	return aca.cfg.AWSRegion
+}
+
+func (aca agentConfigAccessor) Cluster() string {
+	return aca.cfg.Cluster
+}
+
+func (aca agentConfigAccessor) DefaultClusterName() string {
+	return DefaultClusterName
+}
+
+func (aca agentConfigAccessor) External() bool {
+	return aca.cfg.External.Enabled()
+}
+
+func (aca agentConfigAccessor) InstanceAttributes() map[string]string {
+	return aca.cfg.InstanceAttributes
```

Comment:
Should we return a copy of data structures instead of a reference to prevent user of this interface from modifying the Agent's view of the config? Or do we trust the users?

Reply:
Synced up offline, we would trust the users. We could make deep copies of data structures just as a fail-safe, but this comes at a cost (i.e., methods that return data structures could run at O(N) instead of O(1)). Amogh just wanted to raise the fact that any updates made by the user of the interface will reflect on the Agent’s view of the config. We acknowledge this and will intentionally leave as-is given the time complexity tradeoff and that we trust users.